custom_connectors/custom_auth/allocadia.rb

```diff
@@ -0,0 +1,1056 @@
+           }
+         end
+       }
+      end,
+
+      input_fields: lambda do |object_definitions|
+        object_definitions['budget'].only('budgetId').concat(object_definitions['column'].only('columnId')).required('budgetId','columnId')
+      end,
+
+      output_fields: lambda do |object_definitions|
+        [{
+          name: 'choices',
+          type: 'array',
+          of: 'object',
+          properties: object_definitions['choice']
+        }]
+      end,
+
+    },
+
+    get_choices_by_column_name: {
+      description: "Get <span class='provider'>choices by column name</span> " \
+        "in <span class='provider'>Allocadia</span>",
+
+      execute: lambda do |_connection, input|
+       
+        allColumnsObj = []
+        
+        locationFilter = input['location'] ? "location eq \"#{input['location']}\" and " : ""
+        
+        column_names = input['columnNames']&.map { |item| item['columnName'] } || []
+        column_names.each do |columnName|
+          columnObj = get("/v1/budgets/#{input['budgetId']}/columns?$filter=#{locationFilter}name eq \"#{columnName}\"").first || error("Column #{columnName} not found")
+        
+          raw_choices = get("/v1/budgets/#{input['budgetId']}/columns/#{columnObj['id']}/choices")
+          choices = raw_choices&.map do |choice|
+            {
+              choiceId: "#{choice['id']}",
+              label: "#{choice['label']}",
+              externalId: "#{choice['externalAssociations'].size > 0 ? choice['externalAssociations'][0]['externalId'] : nil}"
+            }
+            end
+     
+          columnObj['choices'] = choices
+          allColumnsObj << columnObj
+        end 
+        
+        {
+          columns: allColumnsObj
+        }
+      end,
+
+      input_fields: lambda do |object_definitions|
+        object_definitions['budget'].only('budgetId').
+          concat([
+            {
+              name: 'columnNames',
+              type: :array,
+              of: :object,
+              properties: [{ name: 'columnName' }]
+            }
+          ]).concat(object_definitions['column'].only('location'))
+      end,
+
+      output_fields: lambda do |object_definitions|
+        [{
+           name: 'columns',
+           type: 'array',
+           of: 'object',
+           properties: object_definitions['column']
+        }]
+      end,
+
+    },
+  
+    add_choice: {
+      description: "Add <span class='provider'>column choice</span> " \
+        "in <span class='provider'>Allocadia</span>",
+
+      execute: lambda do |_connection, input|
+
+        param = '{ "label":"' + input['label'] + '", "externalAssociations": [{ "externalId": "' + input['externalId'] + '", "type": "CAMPAIGN" } ] }'
+
+        post("/v1/budgets/#{input['budgetId']}/columns/#{input['columnId']}/choices", parse_json(param))
+          .after_response do |code, body, headers|
```

Comment:
I've partially got this working but want to use "only" to limit the input to label, externalId and type. Since externalId and type are nested, how do I specify those fields in the "only" list? Also, is there a way to configure the default for 'type' to be 'CAMPAIGN' since for now that's the only valid option?

Reply:
I got this working using a simplified object definition with just what's needed for creation. Also found that 'default' in the object definition works (though I didn't see it documented anywhere?)